### ISSUE TITLE
UnitTest needs better output handling / reporting

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3365]
The output of UnitTest is not wonderful when compared to the output of test frameworks such as [`rspec`](http://rspec.info/).  For a start, it's far too verbose.  Generally developers only want to see the details of failures.  If a test passes, traditionally the framework will output nothing more than a single `.` character`.

So one easy improvement would be to change the default reporting behaviour of individual tests so that by default the details of each test (or at least the output from any assertions) is only shown when the test fails.  This could be done by making `report` default to `nil` rather than `true`, where `nil` means "let the test framework choose whether to report", and its choice would be to only output details when there is a failure.

If this sounds like an appealing proposal, I can submit a PR for it.  I already started making the code changes, and they're pretty trivial.

Of course, in the longer run, we could refactor and extract out separate `TestReporter` and `TestFormatter` classes, so that multiple output formats could be supported.

**Comment by snappizz** [comment]
yeah, wholly agree. there is a deeper problem here -- a lot of tests produce insuppressible post output on the side, in particular the server boot ones. it would be an ambitious project, but it's a generally go idea to work on cleaning up all the crap that sclang posts in general.

**Comment by jamshark70** [comment]
> ... a lot of tests produce insuppressible post output on the side, in particular the server boot ones.

Did anyone try setting the verbosity level in the ServerOptions object(s) used in the server tests?

**Comment by aspiers** [comment]
@snappizz Thanks for the reply. Any thoughts on my proposal of adding support for `report: nil`?  This would handle all output from assertions, which I suspect is the majority of output in many tests, although of course it wouldn't handle other output.

Another more comprehensive approach taken by other test frameworks in other languages is to substitute the stdout/stderr stream objects for a stream-like object which captures input into a string.  That way, none of the test code which generates output to stdout or stderr needs to be changed, but the framework can capture the output and decide afterwards whether it should be displayed to the user.

For example, in Ruby, it's the [`StringIO`](https://ruby-doc.org/stdlib-2.5.0/libdoc/stringio/rdoc/StringIO.html) class.  Is there any equivalent in SuperCollider?

**Comment by telephon** [comment]
`CollStream` is close, may need to add more to it

**Comment by telephon** [comment]
the main steps will be: 
- change the `assert` etc. methods to post to a stream object (`Post` by default)
- replace all ` postln` in the unit tests by a different redirected method, like `postd`.

**Comment by telephon** [comment]
of course, it would also be interesting to be able to redirect all sclang print output from a single place. Not difficult, but needs discussion ...

**Comment by aspiers** [comment]
OK, I've made a first small step towards this in #3569.
